README


Black Flower CAD v1.13.2 - 12 AUG 2022

OVERVIEW

  Description:
  ------------
  Black Flower CAD is an open source Computer Aided Dispatch / Computer Aided Logging software package.  
  
  Contact Information:
  --------------------
  The current version of this software can be downloaded from this GitHub or hosted from https://cad.w2ort.com
  
  For bug reports, questions, etc, send email to:
    w2ort.matt@gmail.com


INSTALLATION
  
  Server Requirements:
  --------------------
  * MySQL database server v5.0 or greater
  * PHP version 7+, with MySQL integration.
  * Apache 2.0+ webserver running in Prefork mode, with PHP integration.
  
  * Linux (Debian/CentOS) recommended for mission critical installations.
  * Microsoft Windows (7/10) has also been used in development environments.
    -> IMPORTANT:  Production use on Microsoft Windows is NOT recommended,
       due to limitations of PHP running under Windows-based webservers and
       using Windows' Event Log emulation for syslog.
  
  
  Client Requirements:
  --------------------
  * 1024x768 (or greater) screen resolution required.
    -> 1280x1024 (or greater) resolution is recommended for best results.

  Browser Support:
  -> Chrome: Working
  -> Microsoft Edge: Working
  -> Safari: Untested
  -> Firefox: Untested
  -> Internet Explorer is NOT supported, any other browsers not listed will not be supported.

  * Client system time must be synchronized with CAD server system time.
  
  * Microsoft Windows (7 or higher) is recommended for client operating system.
    -> Linux clients have also been tested successfully.

  
  
  Installation Quick Guide
  ------------------------

BEING WORKED ON
BEING WORKED ON
BEING WORKED ON




  To use the CAD system:
  ----------------------
  On the server, start or restart Apache and MySQL as needed.
  On the client, load your web browser and go to the CAD URL, e.g.:

    http://<server>/<path-to-CAD-application>/

  (or https:// if you have configured your system for HTTPS.)
  
  Log in with the CAD administrator username and password created 
  during step 4 above.  The password is case sensitive.
  
  Click on the "Settings" tab, then "Edit Users" in the Administration 
  dialog, and create the required usernames and passwords for CAD users.

  An access level less than 10 should be used for normal users.  
  Access levels of 10 or greater create a system administrator account.  

  Day to day use of CAD should be done through normal, rather than
  system administrator, accounts.
  
  
  Best Common Practices
  ---------------------
  1)  NTP
  
  Time synchronization is required for coordination of timestamps 
  generated on the server with those generated on the clients.  
  This can be done by running a standalone (stratum 1) NTP server 
  if on a completely isolated network, or by connecting all systems 
  to the Internet NTP network.  In the Internet case it is 
  recommended to run a caching NTP server on the CAD server, and 
  synchronize all clients to that.  A popular freeware Windows NTP 
  client is available from http://www.oneguycoding.com/automachron/.
  
  2)  Syslog
  
  CAD will use the native system logging functionality on UNIX or 
  Windows systems.  On UNIX systems, CAD uses the local4 logging level,
  and obeys the severity filter as set in cad.conf.  On Windows, 
  *all* messages are logged to the Application section of the Event 
  Viewer, disregarding the severity filter as set in cad.conf.  
  
  3)  Network security
  
  Any network running CAD or other such systems should be kept 
  isolated from insecure network traffic such as the Internet.  
  The CAD server should be reachable from the CAD clients but they
  should not be reachable from untrusted (outside/Internet) systems.
  Any consumer-grade or better hardware firewall is sufficient for 
  purposes of CAD network security.  
  
  In an extremely small or field-deployable CAD installation, it is 
  possible to install the server on a system which will also be one
  of the clients of the installation.  If that is the only system 
  expected to use CAD, a software firewall on that host is sufficient 
  if configured correctly.  Software firewalls are also generally 
  recommended for any Microsoft Windows client or server systems 
  as best practice.  CAD is a cleanly client-server system, it does 
  not require network ingress access for connections _to_ clients.
  
  Ports required to be accessible on CAD server by the CAD clients
  for application communication:
  
    80 (http)   - Normal operation: HTTP on a secure network.
  *or*
   443 (https)  - If using secure HTTPS on an insecure network
  
  Other ports that may be used in a full network environment:
  
    53 (domain) - If using the CAD server as a DNS server.
    67 (bootps) - If using the CAD server as a DHCP server.
   123 (ntp)    - If using the CAD server as an NTP server.
  
  
  


CHANGES
CHANGES
-------


---------------------------
Version 1.13.2 - 2022-08-12
---------------------------
[major] Complete overhaul of CAD to PHP v8
[bugfix] Fixed staging location not populating in logs
[bugfix] Fixed log viewer to default to 25
[bugfix] Can now leave field blank when inputing logs
[minor]  Added tone following creation of new call


---------------------------
Version 1.12.3 - 2016-07-22
---------------------------
[bugfix] Fixed Log Viewer from not displaying any messages under certain
         conditions.
[bugfix] Improved the logic for compiling partial incident header 
         information into unit dispatch auto-pages.

----------------------------------------
Version 1.12.2 (unreleased) - 2015-09-01
----------------------------------------
[minor]  On staging screen, show the times when each unit was staged there.

---------------------------
Version 1.12.1 - 2015-08-31
---------------------------
[bugfix] Prevent multiple operators from releasing an attached unit 
         multiple times and by doing so toggling its status irrevocably 
         back 'Attached To Incident'.

---------------------------
Version 1.12.0 - 2015-08-31
---------------------------
[minor]  Added initial version of unit utilization reports.
[minor]  Upgraded to FPDF 1.7, with assorted changes to AddPage() and 
         Output() calls.
[bugfix] Cleaned up various issues with reports.
[minor]  Added unit staging functionality.

---------------------------
Version 1.11.0 - 2014-08-23
---------------------------

[bugfix] Fixed erroneous incident "Undispatched" status that could be 
         displayed if User A created an incident, User B took over the
         incident's read/write lock, but User A then attached a unit 
         prior to refreshing the window and losing their original read/write 
         lock to User B.  Resolved multiple related bug reports.
[minor]  Changed Incident window to display times as full date + time rather
         than just name of day + time, if a timestamp was from a different 
         calendar day.  Display is not very elegant currently, but is more 
         correct functionality.
[minor]  Implemented configurable permissions levels for administrative
         and supervisory functions.
[minor]  Implemented single-incident report.
[minor]  Restricted incident note entry from allowing browser form history.
[minor]  Tweaked incident list "Undispatched" flag display style for increased 
         visibility. (blink property doesn't work everywhere any more.)
[minor]  Fixed config-cleardb to work with Mysql 5.x constraints.

---------------------------
Version 1.10.1 - 2013-09-30
---------------------------

[minor]  Applied viewport setting to stop iPad client from continually 
         refreshing incidents pane bigger than one screen.
[minor]  Added unit response time reports.
[minor]  Enhanced unit reports by allowing range of start and end dates.
[minor]  Added unit daily summary reports.

---------------------------
Version 1.10.0 - 2013-07-18
---------------------------

[minor]  Allow filtering incident reports by incident type.
[minor]  Added ability to prevent closing (dispositioning) incidents unless 
         all assigned units are first manually released from it.  This is 
         an administrator-configurable feature:  in cad.conf, set:
           $FORCE_MANUAL_UNIT_RELEASE= 1;
[minor]  Added ability to require supervisors to review and approve the 
         details of a Dispositioned incident before it is fully Completed.
         This is an administrator-configurable feature:  in cad.conf, set:
           $SUPERVISOR_INCIDENT_REVIEW = 1;
[minor]  Now require cad.conf "$ALLOW_MANUAL_PAGE = 1;" to enter a page 
         to a unit manually from CAD.  Automatic page-out still works 
         regardless of this setting.
[bugfix] Disable channel assignment functionality on a closed incident.
[bugfix] Disable note "submit" button on a closed incident.


--------------------------
Version 1.9.2 - 2013-04-17
--------------------------

[minor]  Importing minor additions for icon definitions and preloaded 
         incident type 'COURTESY TRANSPORT', changes were made but not 
         committed in August.

--------------------------
Version 1.9.1 - 2012-08-25
--------------------------

[minor]  Added ability to resolve a call as a duplicate of another call.

[minor]  Don't show duplicate/training calls in status totals.

[minor]  Focus cursor to unit name field on new unit screen.

[minor]  Improved cosmetics of incidents reports and unit reports.

--------------------------
Version 1.9.0 - 2012-08-21
--------------------------
[major]  THE PASSWORD HASH FORMAT HAS CHANGED AS OF 1.9.0.  All passwords 
         generated on previous versions must be reset per this example
         (starting with the Administrator account for convenience):

  1. Set $DEBUG=1 in cad.conf, 
  2. Attempt to log in as Administrator (which will fail), 
  3. Observe the hash displayed on the login failure screen,
  4. Update the MySQL users table with the displayed hash: 
     mysql>  update cad.users set password='<hash>' where id=1;
  5. Log in as Administrator with your new password.
  6. Go to the System Admin tab and change all other users' passwords.
  7. Set $DEBUG=0 in cad.conf.

[major]  Implemented channel assignments to incidents.  Channels may
         be configured by the sysadmin to be available or unavailable
         (for reference only), repeater or simplex (for display purposes),
         then assigned by a dispatcher in the edit_incident window.

[major]  Implemented optional read-only mode and associated config settings.
         If read-only settings are not defined, the app will remain wide open
         (in read-write mode) for all connections.
         

cad.php
<?php
  $subsys="cad";

  require_once('db-open.php');
  require_once('session.inc');
  require_once('functions.php');

  if (isset($_POST['unit']))  {
    SessionErrorIfReadonly();
    $post_pend="";
    $unit = MysqlClean($_POST, 'unit', 20);

    if (isset($_POST['message']))
      $message = MysqlClean($_POST, 'message', 255);
    else
      die("Internal error: POST message value missing.");

    if (isset($_SESSION['username']) && $_SESSION['username'] != '') {
      $creator = $_SESSION['username'];
    }
    else {
      $creator = '';
    }

    if (isset($_POST['message_type']))
      $message_type = MysqlClean($_POST, 'message_type', 20);
    else
      $message_type = "";

    $query = "INSERT INTO messages (ts,unit,message,creator,message_type) VALUES (NOW(), UPPER('$unit'), '$message', '$creator', '$message_type')";
    MysqlQuery($query);

    if (isset($_POST["hour"]))
      $post_pend .= "&hour=". MysqlClean($_POST, "hour", 20);
    # TODO: is this code still used?  there should also be a 'date' in post_pend.

    header("Location: cad.php$post_pend");
  }

  header_html("Dispatch :: Log Viewer","  <script src=\"js/clock.js\" type=\"text/javascript\"></script>");
?>
<body vlink="blue" link="blue" alink="cyan" onload="displayClockStart()" onunload="displayClockStop()" onresize="resizeMe()">
<?php include('include-title.php'); ?>

<!-- Begin Add Message Form -->
<form name="myform" action="cad.php" method="post" style="margin: 0px;">

<b class="text">Log Viewer
<?php if (!isset($_SESSION['readonly']) || !$_SESSION['readonly']) { print ':: Add a Message:'; } ?>
</b><br>


<table width="100%" style="margin-bottom: 8px;">
<?php if (!isset($_SESSION['readonly']) || !$_SESSION['readonly']) { ?>
<tr>
<td bgcolor="#aaaaaa">

<table width="100%" cellspacing="1" cellpadding="2">
<tr valign="bottom" style='font-size: 8pt;' class='message'>
   <td class="text" nowrap>Message time:</td>
   <td class="text" nowrap><u>S</u>elect unit from list:&nbsp;</td>
 <?php
   if ($USE_MESSAGE_TYPE) {
     if (!isset($_COOKIE['cad_show_message_type']) || $_COOKIE['cad_show_message_type'] == 'yes') {
       print '<td class="text" nowrap>Message type:&nbsp;</td>';
     }
  }
 ?>
   <td class="text">Enter <u>m</u>essage</td>
   <td nowrap rowspan="2">
     <input type="submit" value="" style="visibility: hidden; width: 0px; height: 0px;"><br>
     &nbsp;&crarr;&nbsp;
     <input type="reset" value="Clear" tabindex="5">
   </td>
</tr>
<?php } ?>
<tr class="message">
   <td><input type="text" name="displayClock" size="8" /></td>
   <td>
     <label for="unit" accesskey="s">
     <select name="unit" id="unit" tabindex="3" style="width:150px">
     <option selected value=""></option>
<?php
      $unitquery = "SELECT unit FROM units";
      $unitresult = MysqlQuery($unitquery) or die("In query: $unitquery<br />\nError: " . mysql_error());
      $unitnames = array();
      while ($line = mysqli_fetch_array($unitresult, MYSQLI_ASSOC)) {
        array_push($unitnames, $line["unit"]);
      }
      natsort($unitnames);
      foreach ($unitnames as $unitname) {
         echo "<option value=\"$unitname\">$unitname</option>\n";
      }
      mysqli_free_result($unitresult);
 ?>
     </select>
     </label>
   </td>

 <?php
    if ($USE_MESSAGE_TYPE) {
      if (!isset($_COOKIE['cad_show_message_type']) || $_COOKIE['cad_show_message_type'] == 'yes') {
        print '<td>';
        print '<select tabindex="4" style="width:150px" name="message_type">';
        print '<option selected value=""></option>';
        $typequery = "SELECT message_type FROM message_types ORDER BY message_type ASC";
        $typeresult = MysqlQuery($typequery) or die("In query: $typequery<br />\nError: " . mysql_error());
        while ($line = mysqli_fetch_array($typeresult, MYSQLI_ASSOC)) {
           print "<option value=\"". $line["message_type"] ."\">". $line["message_type"] ."</option>\n";
        }
        mysqli_free_result($typeresult);
        print "</select>\n</td>\n";
      }
    }
 ?>

   <td>
   <label for="message" accesskey="m">
   <input type="text" name="message" id="message" tabindex="4" size="70" maxlength="250" />
   </label>
   </td>
</tr>
</table>

</td>
</tr>
</table>

</form>
<!-- End Add Message Form -->

<!-- Begin Filter Form -->
<form name="filter" action="cad-log-frame.php" method="post" style="margin: 0px;" target="log">

<span class="text"><b style="font-size: 10pt;">Log Viewer :: Viewing messages</b>
&nbsp; Use selections below to filter for specific log messages.
</span>

<!-- Begin Filter Form Outer Table -->
<table width="100%" style="margin-bottom: 8px;">
<tr>
<td bgcolor="#aaaaaa">

  <!-- Begin Filter Form Inner Table -->
  <table width="100%" cellspacing="1" cellpadding="2">
  <tr class="message">

  <td class="text">Date:&nbsp;</td>
  <td class="text">
  <select name="date" id="date" tabindex="101">
  <option value=""></option>
<?php
  $datesquery = "SELECT DISTINCT CAST(ts AS DATE) AS tsdate FROM messages ORDER BY ts DESC";
  $datesresult = MysqlQuery($datesquery);
  $dates = array();
  while ($line = mysqli_fetch_array($datesresult, MYSQLI_ASSOC)) {
    array_push($dates, $line["tsdate"]);
  }
  foreach ($dates as $date) {
    echo "<option value=\"$date\">$date</option>\n";
  }
  mysqli_free_result($datesresult);
?>
  </select>
  </td>

  <td class="text">Hour:&nbsp;</td>
  <td class="text">
  <select name="hour" id="hour" tabindex="102">
  <option value=""></option>
<?php
  for ($i = 0; $i < 24; $i++) {
    echo "<option value=\"$i\">";
    if ($i < 10) echo "0";
    print "$i:00</option>\n";
  }
?>
  </select>
  </td>

  <td class="text">Unit:&nbsp;</td>
  <td class="text">
  <select name="funit" id="funit" tabindex="103">
  <option value=""></option>
<?php
  $unitquery = "SELECT unit FROM units";
  $unitresult = MysqlQuery($unitquery);
  $unitnames = array();
  while ($line = mysqli_fetch_array($unitresult, MYSQLI_ASSOC)) {
    array_push($unitnames, $line["unit"]);
  }
  natsort($unitnames);
  foreach ($unitnames as $unitname) {
    echo "<option value=\"$unitname\">$unitname</option>\n";
  }
  mysqli_free_result($unitresult);
?>
  </select>
  </td>

  <td class="text" nowrap>Messages Per Page:&nbsp;</td>
  <td class="text">
  <select name="mpp" id="mpp" tabindex="104">
<?php
  $mpp = array(10, 25, 100);

  // This should probably be abstracted to a configurable setting at some point...
  $mppdefault = 25;

  $filtermppindex = 0;
  $filtermppcount = 0;

  echo "<option value=\"0\">All</option>\n";
  $filtermppcount++;

  foreach ($mpp as $pp) {
    echo "<option value=\"$pp\"";
    if (isset($mppdefault) && $mppdefault == $pp) {
      echo " SELECTED";
    }
    echo ">$pp";
    if (isset($mppdefault) && $mppdefault == $pp) {
      echo " (Default)";
      $filtermppindex = $filtermppcount;
    }
    echo "</option>\n";
    $filtermppcount++;
  }
?>
  </select>
<?php
  echo "<SCRIPT LANGUAGE=\"JavaScript\">var filtermppindex = $filtermppindex;</SCRIPT>\n";
?>

  </td>

  <td class="text" nowrap align="right" width="100%">
  <button type="submit" name="apply_filters" id="apply_filters" value="apply_filters">Filter</button>
  <button type="submit" name="remove_filters" id="remove_filter" value="remove_filters"
   onClick="document.getElementById('date').options[0].selected=true;
            document.getElementById('hour').options[0].selected=true;
            document.getElementById('funit').options[0].selected=true;
            document.getElementById('mpp').options[filtermppindex].selected=true;"
   >Reset</button>
  </td>

  </tr>
  </table>
  <!-- End Filter Form Inner Table -->

</td>
</tr>
</table>
<!-- End Filter Form Outer Table -->

</form>
<!-- End Filter Form -->

<iframe name="log" src="cad-log-frame.php"
        width=<?php print trim($_COOKIE['width'])-30; ?>
        height=<?php print trim($_COOKIE['height'])-250; ?>
        marginheight="0" marginwidth="0" frameborder="0"> </iframe>
<?php
   mysqli_close($link);
?>
</body>
</html>
